benchmark-ganache.ts

```diff
@@ -0,0 +1,165 @@
+];
+
+main();
+
+async function main(): Promise<void> {
+  /*
+   * setup
+   */
+
+  console.log(chalk.bold("Setting up Ganache..."));
+  console.time("setup-ganache");
+  console.group();
+
+  const { ganache, provider } = await prepareGanache({
+    url,
+    blockNumber,
+    blockGasLimit,
+    defaultBalance,
+    deleteCache: false
+  });
+
+  const convex = new ethers.Contract(convexAddress, convexAbi, provider);
+  const ownerAddress = await convex.owner();
+
+  await fundAccounts({
+    provider,
+    accounts: [convexAddress, ownerAddress],
+    amount: targetBalance
+  });
+
+  const owner = await unlockAddress({
+    provider,
+    address: ownerAddress
+  });
+
+  console.groupEnd();
+  console.timeEnd("setup-ganache");
+  console.log("");
+
+  /*
+   * simulation
+   */
+
+  console.log(chalk.bold("Simulating shutdown..."));
+  console.time("simulate-shutdown");
+  console.group();
+  const tx = await convex.connect(owner).shutdownSystem({
+    gasLimit: "0x25317C0" // 39,000,000
+  });
+
+  const receipt = await provider.waitForTransaction(tx.hash);
+  console.groupEnd();
+  console.timeEnd("simulate-shutdown");
+  console.log("");
+
+  // @ts-ignore looks like ganche has a bad typing
+  await ganache.disconnect();
+}
+
+interface PrepareGanacheOptions {
+  url: string;
+  blockNumber: "latest" | number;
+  blockGasLimit: string;
+  defaultBalance: string;
+  deleteCache: boolean;
+}
+
+async function prepareGanache({
+  url,
+  blockNumber,
+  blockGasLimit,
+  defaultBalance,
+  deleteCache
+}: PrepareGanacheOptions): Promise<{
+  ganache: ReturnType<typeof Ganache.provider>;
+  provider: ethers.providers.JsonRpcProvider;
+}> {
+  const ganache = Ganache.provider({
+    fork: {
+      url,
+      blockNumber,
+      deleteCache
+    },
+    miner: { blockGasLimit },
+    wallet: {
+      // ganache expects value in ETH
+      defaultBalance: ethers.utils.formatEther(defaultBalance)
+    },
+    logging: {
+      quiet: true
+    },
+    legacyInstamine: true
+  });
+
+  const provider = new ethers.providers.Web3Provider(ganache);
+
+  return { ganache, provider };
+}
+
+interface FundAccountsOptions {
+  provider: ethers.providers.JsonRpcProvider;
+  accounts: string[];
+  amount: string;
+}
+
+async function fundAccounts({
+  provider,
+  accounts,
+  amount
+}: FundAccountsOptions): Promise<void> {
+  const [from] = await provider.listAccounts();
+
+  for (const address of accounts) {
+    // simple contract that just selfdestructs funds to address constructor arg
+    const sendBytecode = "0x60806040526040516100c13803806100c18339818101604052810190602391906098565b8073ffffffffffffffffffffffffffffffffffffffff16ff5b600080fd5b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000606a826041565b9050919050565b6078816061565b8114608257600080fd5b50565b6000815190506092816071565b92915050565b60006020828403121560ab5760aa603c565b5b600060b7848285016085565b9150509291505056fe";
```

Comment:
Does ganache have an RPC method to fund accounts? This works fine of course, again mainly just curious

Reply:
No, although I suspect we'll be adding one soon :)

Reply:
> Note: this leaves out the log decoding work that I mentioned in my tweet because it won't work reliably without an etherscan API key.
> 
> Happy to add that back in or make any other modifications!
> 
> Ganache fork caching doesn't seem to be working using fork.blockNumber `"latest"`

This is actually expected, as state may have changed at any block between the previous `"latest"`'s cached state and the new "latest" block.